### PR TITLE
Arm sensors

### PR DESCRIPTION
**Pull request by greenblitz4590**

### CODE REVIEW

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java [hunk 1]
```diff
@@ -5,6 +5,8 @@
 import edu.greenblitz.tobyDetermined.commands.telescopicArm.extender.ExtendToLength;
 import edu.greenblitz.tobyDetermined.subsystems.telescopicArm.Elbow;
 import edu.greenblitz.tobyDetermined.subsystems.telescopicArm.Extender;
+import edu.wpi.first.math.system.plant.DCMotor;
```
Comment (roygold8) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java:8]: We dont do sim in master

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java [hunk 2]
```diff
@@ -15,9 +17,34 @@ public GoToPosition(double lengthInMeters, double angleInRads) {
         }
         //if the desired position needs to pass through the entrance zone and the extension to long the movement is split to multiple parts
         else{
```
Comment (roygold8) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java:19]: More doc for the function

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java [hunk 3]
```diff
@@ -15,9 +17,34 @@ public GoToPosition(double lengthInMeters, double angleInRads) {
...
+                                .alongWith(new RotateToAngle(RobotMap.telescopicArm.elbow.STARTING_WALL_ZONE_ANGLE))
+                );
+                addCommands(new RotateToAngle(RobotMap.telescopicArm.elbow.END_WALL_ZONE_ANGLE)
+                        .andThen(
+                                new RotateToAngle(angleInRads)
+                                        .andThen(new ExtendToLength(lengthInMeters))
+                        )
+                );
+            }else{
+                // STOPSHIP: 2/15/2023
```
Comment (roygold8) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java:46]: ?

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Claw.java [hunk 4]
```diff
@@ -30,15 +38,37 @@ public static void init(){
...
+    public void motorEject() {
+        motor.set(ControlMode.PercentOutput,RobotMap.telescopicArm.claw.MOTOR_POWER_RELEASE);
     }
 
     public void stopMotor (){
-        motor.set(0);
+        motor.set(ControlMode.PercentOutput,0);
+    }
+
+    public enum ClawState{
```
Comment (roygold8) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Claw.java:70]: Something to describe if we have an item?

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Elbow.java [hunk 5]
```diff
@@ -73,11 +83,12 @@ public void setAngleByPID(double goalAngle) {
                 pidController.getSetpoint().velocity, (pidController.getSetpoint().velocity - lastSpeed) / RoborioUtils.getCurrentRoborioCycle(),
                 Extender.getInstance().getLength(), getAngle()
         );
+        SmartDashboard.putNumber("Elbow FF", feedForward);  //todo - its for debugging, remove when done
```
Comment (roygold8) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Elbow.java:86]: I think we can remove? we have full debug dashboard?

**Review by roygold8 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java [hunk 6]
```diff
@@ -10,140 +11,160 @@
...
+		return instance;
+	}
+	
+	public static void init() {
+		instance = new Extender();
+	}
+	
+	private Extender() {
+		motor = new GBSparkMax(RobotMap.telescopicArm.extender.MOTOR_ID, CANSparkMaxLowLevel.MotorType.kBrushless);
+		motor.getEncoder().setPosition(0);
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java:36]: move to const - there is an option that the extender will not start at 0

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java [hunk 7]
```diff
@@ -10,140 +11,160 @@
...
+				RobotMap.telescopicArm.extender.kV * wantedSpeed +
+				RobotMap.telescopicArm.extender.kA * wantedAcceleration;
+	}
+	
+	public static double getStaticFeedForward(double elbowAngle) {
+		return Math.sin(elbowAngle - RobotMap.telescopicArm.elbow.STARTING_ANGLE_RELATIVE_TO_GROUND) * RobotMap.telescopicArm.extender.kG;
+	}
+	
+	private void setLengthByPID(double lengthInMeters) {
+		profiledPIDController.reset(getLength());
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java:83]: reset to state (like in movetopose)

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java [hunk 8]
```diff
@@ -10,140 +11,160 @@
...
+	private boolean lastSwitchReading;
+	
+	/**
+	 * @return whether the switch changed status from last periodic this is useful to reset at
+	 */
+	public boolean didSwitchFlip(){
+		return lastSwitchReading ^ getLimitSwitch();
+	}
+	
+	public void resetLength() {
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java:134]: make reset to receive an angel and default to 0

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java [hunk 9]
```diff
@@ -10,140 +11,160 @@
...
+		IN_ROBOT_BELLY_LENGTH,
+		//the state corresponding to OUT_ROBOT
+		OPEN,
+		//the state corresponding to WALL_ZONE
+		IN_WALL_LENGTH,
+		// this state should not be possible and is either used to stop dangerous movement or to signal a bug
+		OUT_OF_BOUNDS
+	}
+	
+	public boolean isAtLength(double wantedLengthInMeters) {
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java:154]: profiledPIDController.atGoal()?

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Claw.java [hunk 10]
```diff
@@ -1,27 +1,35 @@
...
+     * forward -> cone
+     * reverse -> cube
+     *
+     * */
 
     public static Claw getInstance() {
         if (instance == null) {
             init();
-            SmartDashboard.putBoolean("claw initialized via getinstance", true);
+            SmartDashboard.putBoolean("claw initialized via getInstance", true);
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Claw.java:32]: I think this should be removed

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/Dashboard.java [hunk 11]
```diff
@@ -44,4 +54,19 @@ public void periodic() {
 		SmartDashboard.putNumber("grid pos id", Grid.getInstance().getSelectedPositionID());
 		SmartDashboard.putString("grid pos", Grid.getInstance().getSelectedPosition().toString());
     }
+	
+	public void debugArm(){
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/Dashboard.java:58]: delete - romi is working on an entire debug dashboard

**Review by yonnbro — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Claw.java (hunk 4 above)
Comment (NoamRosenberg08) [reply src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Claw.java:70]: we don't know if we will have a switch there

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java (hunk 8 above)
Comment (NoamRosenberg08) [reply src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java:134]: angle?

Comment (roygold8) [reply src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java:134]: length in Extender, angle in Elbow

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java (hunk 2 above)
Comment (roygold8) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java:19]: You need to add doc for those if statements

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Claw.java (hunk 4 above)
Comment (roygold8) [reply src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Claw.java:70]: I think, in the case, we don't - we should go the naive way and say - if we gripped already, we have an item - those states should be useful.

@yonnbro  what do you think?

**Review by roygold8 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java (hunk 2 above)
Comment (roygold8) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java:20]: MORE :)

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/Dashboard.java [hunk 12]
```diff
@@ -20,11 +24,17 @@ public static void init(){
...
+	
+	
     @Override
     public void periodic() {
+	
 
-        SmartDashboard.putNumber("FR-angle-neo", Math.IEEEremainder(Math.toDegrees(SwerveChassis.getInstance().getModuleAngle(SwerveChassis.Module.FRONT_RIGHT)), 360));
+	
+	
+
```
Comment (roygold8) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/Dashboard.java:36]: Dont merge dashboard changes

**Review by roygold8 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/tobyDetermined/RobotMap.java [hunk 13]
```diff
@@ -246,7 +246,7 @@ public static class extender {
             public static final double GEAR_RATIO = 1;
             public static final double EXTENDED_LENGTH = 0.6;
             public static final double SHRINKED_LENGTH = 0.6;
-
+            public static final int STARTING_POS = 0;
```
Comment (greenblitz4590) [inline src/main/java/edu/greenblitz/tobyDetermined/RobotMap.java:249]: Raz - should be double, use POSITION insteaf of POS

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java [hunk 14]
```diff
@@ -10,14 +10,41 @@
 public class GoToPosition extends SequentialCommandGroup {
 
     public GoToPosition(double lengthInMeters, double angleInRads) {
+        //if the extender is inside the robot.
         if(Extender.getHypotheticalState(lengthInMeters) == Extender.ExtenderState.IN_WALL_LENGTH || Elbow.getInstance().isInTheSameState(angleInRads)){
```
Comment (greenblitz4590) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/GoToPosition.java:14]: Raz - This is a realllllly long line

File: src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java [hunk 15]
```diff
@@ -10,140 +11,165 @@
...
+	public double getVelocity(){return motor.getEncoder().getVelocity();}
+	
+	/**
+	 * @return the current value of the limit switch
+	 */
+	public boolean getLimitSwitch() {
+		return motor.getReverseLimitSwitch(SparkMaxLimitSwitch.Type.kNormallyOpen).isPressed();
+	}
+	
+	private boolean lastSwitchReading;
```
Comment (greenblitz4590) [inline src/main/java/edu/greenblitz/tobyDetermined/subsystems/telescopicArm/Extender.java:126]: Raz - move this to the begging of the class

**Pull request merged**